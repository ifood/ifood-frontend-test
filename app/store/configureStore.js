import createSagaMiddleware from 'redux-saga';
import { compose, createStore as createReduxStore, applyMiddleware } from 'redux';
import memoizee from 'memoizee';
import createReducers from './reducers';
import subscribeStore from './subscribeStore';
import { loadFromLocalStorage } from './localStorage';

const sagaMiddleware = createSagaMiddleware();

// we need to memoize saga to prevent duplicated entries
function runSaga(saga) {
  sagaMiddleware.run(saga);
}

const sageMemoized = memoizee(runSaga);

// we need to memoize reducers to prevent duplicated entries
function registerReducer(name, reducer, store, lazyReducers) {
  if (typeof name !== 'string') {
    throw new TypeError('"name" is required and must be a string.');
  }

  if (typeof reducer !== 'function') {
    throw new TypeError('"reducer" is required and must be a reducer function.');
  }

  if (store === undefined || store === null) {
    throw new TypeError('"store" is required and must not be null or undefined');
  }

  lazyReducers[name] = reducer;

  store.replaceReducer(createReducers(lazyReducers));

  return store;
}

const reduxMemoized = memoizee(registerReducer);

// redux debug
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default function configureStore() {
  const lazyReducers = {};
  // gets persisted values from localStorage
  const preloadedState = loadFromLocalStorage();
  const store = createReduxStore(
    createReducers(),
    preloadedState,
    composeEnhancers(applyMiddleware(sagaMiddleware)),
  );

  // allow splitted sagas to run from other parts of the app besides here
  store.runSaga = (saga) => sageMemoized(saga);

  // global sagas
  // store.runSaga(SomeSaga);

  // global subscribers, updates localStorage
  subscribeStore(store);

  // for lazy/splitted reducers
  store.registerReducer = (name, reducer) =>
    reduxMemoized(name, reducer, store, lazyReducers);

  return store;
}
