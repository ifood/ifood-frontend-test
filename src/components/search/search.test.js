import React from 'react'
import { Provider } from 'react-redux'
import { render } from '@testing-library/react'
import configureStore from 'redux-mock-store'
import { Search } from '.'

import filtersMock from '../../__mocks__/filters.json'

const mockStore = configureStore([])

describe('Search', () => {
  let store
  let component

  beforeEach(() => {
    store = mockStore({
      filters: {
        filters: []
      },
      playlists: {
        error: null
      }
    })

    component = render(
      <Provider store={store}>
        <Search filters={filtersMock.filters} />
      </Provider>
    )
  })

  it('should render component correctly', () => {
    const { asFragment } = component
    expect(asFragment()).toMatchSnapshot()
  })

  it('should find title text', () => {
    const { getByText } = component
    const linkElement = getByText(/Spotifood/i)
    expect(linkElement).toBeInTheDocument()
  })
})
