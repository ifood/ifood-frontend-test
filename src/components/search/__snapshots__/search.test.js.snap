// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render component correctly 1`] = `
<DocumentFragment>
  <h1
    class="MuiTypography-root makeStyles-title-1 MuiTypography-h1 MuiTypography-gutterBottom"
  >
    Spotifood
  </h1>
  <form
    novalidate=""
  >
    <div
      class="MuiFormControl-root makeStyles-formControl-33"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
      >
        Locale
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="field-locale field-locale"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="field-locale"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          type="hidden"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiFormControl-root makeStyles-formControl-33"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
      >
        País
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="field-country field-country"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="field-country"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          type="hidden"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiFormControl-root makeStyles-formControl-115"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="field-timestamp"
          id="field-timestamp-label"
        >
          Data e Horário
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            id="field-timestamp"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="makeStyles-formControl-155"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
      >
        Quantidade
      </label>
      <span
        class="MuiSlider-root MuiSlider-colorPrimary"
        id="field-limit"
      >
        <span
          class="MuiSlider-rail"
        />
        <span
          class="MuiSlider-track"
          style="left: 0%; width: 18.367346938775512%;"
        />
        <input
          name="limit"
          type="hidden"
          value="10"
        />
        <span
          aria-label="Quantidade"
          aria-orientation="horizontal"
          aria-valuemax="50"
          aria-valuemin="1"
          aria-valuenow="10"
          class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-176"
          data-index="0"
          role="slider"
          style="left: 18.367346938775512%;"
          tabindex="0"
        >
          <span
            class="PrivateValueLabel-offset-178 MuiSlider-valueLabel"
          >
            <span
              class="PrivateValueLabel-circle-179"
            >
              <span
                class="PrivateValueLabel-label-180"
              >
                10
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      class="makeStyles-formControl-155"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
      >
        Página
      </label>
      <span
        class="MuiSlider-root MuiSlider-colorPrimary"
        id="field-offset"
      >
        <span
          class="MuiSlider-rail"
        />
        <span
          class="MuiSlider-track"
          style="left: 0%; width: 0%;"
        />
        <input
          name="offset"
          type="hidden"
          value="0"
        />
        <span
          aria-label="Página"
          aria-orientation="horizontal"
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-176"
          data-index="0"
          role="slider"
          style="left: 0%;"
          tabindex="0"
        >
          <span
            class="PrivateValueLabel-offset-178 MuiSlider-valueLabel"
          >
            <span
              class="PrivateValueLabel-circle-179"
            >
              <span
                class="PrivateValueLabel-label-180"
              >
                0
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submitButton-2 MuiButton-containedPrimary"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </span>
          Buscar
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
